### DISCUSSION TITLE
PyTorch + cuQuantum Batched Circuit Initialization

### DISCUSSION THREAD

**Discussion by wcqc** [NVIDIA/cuQuantum#84] [Q&A] [answered]
In the Python API for cuStateVec, there is batched measurements for example, but there isn't a batched version of state vec initialization. Suppose I have a batched PyTorch tensor of state vecs, any guidance on what would be the best way to batch initialize with cuStateVec?

**Comment by mtjrider** [comment, marked as answer]
Hi @wcqc 

You are right. There is no batch API for state initialization. If the target initialization state is simple, and does not require complex index slicing, it may be more performant to use library-native slicing. For example, if we want to initialize to $\ket{00\cdots0}$, this can be done with:

```
import torch
...
vectors = torch.zeros((size, 2**num_qubits),...)
vectors[:, 0] = 1
```

The compiled kernel launch configurations for `custatevec.initialize_state_vector(...)` will usually run with very high thread occupancy, limiting our ability pipeline a sequence of calls. There is a silver lining when index slicing becomes complex, e.g., with the `W` state:

```
...
index = torch.tensor([2**i for i in range(num_qubits)])
vectors[:, index] = 1 / np.sqrt(num_qubits)
```

In this situation, the selection criteria generate undesirable overheads. Because we've written dedicated routines to handle this case, custatevec tends to outperform this index slicing pattern for systems that are large enough. Here's a snippet of a profile I generated on A100 80GB:

![image](https://github.com/NVIDIA/cuQuantum/assets/13078940/d0681ddf-db65-4e62-9bac-29d99a0c78df)

To guarantee that there are no gaps (worst case -- ideally, you'd have overlap), you can introduce multiple streams, and launch multiple initializations. In my example, I chose a batch size of 8 with 8 batches resulting 64 state vectors each with 24 qubits. Here is what the pipeline ended up looking like in the profile:

![image](https://github.com/NVIDIA/cuQuantum/assets/13078940/52bf1ef0-ada9-4676-8182-22c5fb757072)

Note that the default stream is empty. To accomplish this, you have to:

1. Create multiple streams.
2. For each stream, create a custatevec handle.
3. Set the appropriate stream for each handle.
4. Within a stream context, launch state vector initialization.

Like this:
```
...
batch_size = 8
num_batches = 8
size = batch_size * num_batches
streams = [torch.cuda.Stream(device=device) for _ in range(batch_size)]
handles = [cusv.create() for _ in streams]
for i in range(batch_size):
    cusv.set_stream(handles[i], streams[i].cuda_stream)
...
for i in range(size):
    index = i % batch_size
    with torch.cuda.stream(streams[index]):
        cusv.initialize_state_vector(
            handles[index],
            vectors_a[i].data_ptr(),
            cusv_state_dtype,
            num_qubits,
            initial_state_type
        )
...
```

I've created a gist [containing the full example](https://gist.github.com/mtjrider/17c512799e7488e387d32251c18c087d). Depending on your use-case, it may not be worth the increased code complexity. We'll look into a batch API for state vector initialization of many small systems.

Let me know if you have other questions.

**Reply by leofang** [reply]
Let me add that we had a plan to also release a batched sv init API, but due to resource constraints we've deprioritized it. Let us know if you prefer to use our batched API over Matt's solution (which should serve most use cases already, I'd think, as the non-batched sv init API is also new and before that no one came to ask for it 🙂).
